src/core/tsi/ssl_transport_security.c

```diff
@@ -846,6 +858,51 @@ static const tsi_frame_protector_vtable frame_protector_vtable = {
     ssl_protector_destroy,
 };
 
+/* --- tsi_server_handshaker_factory methods implementation. --- */
+
+static void tsi_ssl_handshaker_factory_destroy(
+    tsi_ssl_handshaker_factory *self) {
+  GPR_ASSERT(self != NULL);
+  GPR_ASSERT(self->vtable != NULL);
+  if (self->vtable->destroy != NULL) {
+    self->vtable->destroy(self);
+  }
+  /* Note, we don't free(self) here because this object is always directly
+   * embedded in another object. If tsi_ssl_handshaker_factory_init allocates
+   * any memory, it should be free'd here. */
+}
+
+static void tsi_ssl_handshaker_factory_addref(
+    tsi_ssl_handshaker_factory *self) {
+  GPR_ASSERT(self != NULL);
+  gpr_refn(&self->refcount, 1);
+}
+
+static int tsi_ssl_handshaker_factory_unref(tsi_ssl_handshaker_factory *self) {
+  int last_reference = 0;
+  GPR_ASSERT(self != NULL);
+
+  last_reference = gpr_unref(&self->refcount);
+  if (last_reference) {
+    tsi_ssl_handshaker_factory_destroy(self);
+  }
+
+  return last_reference;
+}
+
+static tsi_ssl_handshaker_factory_vtable handshaker_factory_vtable = {NULL};
+
+/* Initializes a tsi_ssl_handshaker_factory object. Caller is responsible for
+ * allocating memory for the factory. */
+static tsi_result tsi_ssl_handshaker_factory_init(
```

Comment:
shouldn't this return `void` instead? This complicates the error handling on the caller side.

Reply:
Since the caller needs to check the return value regardless of what is returned, returning a `tsi_result` would be more expressive of the underlying error. Do you mind reviewing this again?

Reply:
Why does the caller need to check the return value if there is no return value (void)? I will add a comment at the caller site.